### ISSUE TITLE
LookSAM Optimizer

### ISSUE THREAD

**Issue by rollovd** [mosaicml/composer#3192]
## New Method
Paper [Towards Efficient and Scalable Sharpness-Aware Minimization](https://arxiv.org/pdf/2203.02714.pdf)

Code repo: https://github.com/rollovd/LookSAM?tab=readme-ov-file 

## Motivation
The main importance of this method is an acceleration of existing SAM. Instead of computing the inner gradient ascent every step, LookSAM computes it periodically and reuses the direction that promotes to flat regions.

According to official paper we can find significant improvement on CIFAR100 and ImageNet-1K.
Also in my work this algorithm also works and shows same metrics as SAM, while saving more time during training, which is important (I'm mostly working with image classification tasks).

Since SAM is already in your repository, it would be great to have an accelerated version of this algorithm.

## Attribution
The paper's authors are `{liuyong, siqimai, youy}@comp.nus.edu.sg,{xiangning, chohsieh}@cs.ucla.edu`.
I wrote the code myself and did not collaborate with the authors.

## [Optional] Implementation

Current implementation works in this way (but of course it can be rewritten).

```
from looksam import LookSAM


model = YourModel()
criterion = YourCriterion()
base_optimizer = YourBaseOptimizer
loader = YourLoader()

optimizer = LookSAM(
    k=10,
    alpha=0.7,
    model=model,
    base_optimizer=base_optimizer,
    rho=0.1,
    **kwargs
)

...

model.train()

for train_index, (samples, targets) in enumerate(loader):
    ...

    loss = criterion(model(samples), targets)
    loss.backward()
    optimizer.step(
        t=train_index, 
        samples=samples, 
        targets=targets, 
        zero_sam_grad=True, 
        zero_grad=True
    )
    ...
```

**Comment by mvpatel2000** [comment]
This looks like a cool paper! It's not on our current roadmap, but we'd love a community PR.

Although, given the example code, it seems like it should just work out of the box. Is there something specific you would like to request with this issue?

**Comment by mvpatel2000** [comment]
@rollovd The example code you provided seems like a great solution to me that works with no code changes! Thanks for providing that, hopefully it is useful to the community. 

I will close this issue out as I don't think we need any code changes to Composer to use this new optimizer :)
